### DISCUSSION TITLE
Warping function between a pair of open curves

### DISCUSSION THREAD

**Discussion by kiranvad** [jdtuck/fdasrsf_python#29] [Q&A] [answered]
Hi @jdtuck,
I have a quick question regarding computing the warping function between a pair of open curves.
I am trying to align the solid black curve to the solid blue curve shown below but the result after the optimization (dotted black curve) doesn't move much away from the solid black curve.
![image](https://user-images.githubusercontent.com/48496966/177886354-7d992a08-4678-4a80-9cb4-7a7ebd8b9337.png)

I am doing this using the following code:
```python
beta2best, q2best, Rbest, gamIbest = fs.find_rotation_and_seed_coord(f1, f2, closed=0, rotation=False, method="DP")
```
where `f1, f2` are of the shape (2, 100) and I am plotting `beta2best` as the warped dotted black curve above.

In reality, these can be aligned using the `time_warping` module but that seems to only align the major peak (see the plot below) of the two curves while I am particularly interested in the full diffeomorphism (i.e. the gamma function that optimally matches one curve to other by stretching and bending) that aligns the two curves.

![image](https://user-images.githubusercontent.com/48496966/177886998-0df8398f-e0ad-4f65-9aad-47b193e1ab9a.png)

Do you have any suggestions? 

TIA
Kiran Vaddi

**Comment by jdtuck** [comment, marked as answer]
Those look like 1D functions then open curves in R^2. The time_warping is the correct module and method and is doing exactly as expected. You are aligning two functions which one has two peaks (black) and the other has multiple (blue). What exactly are you expecting? The time_warping aligns the two peaks. The diffeo is the gamma function.

/vr

Derek



J. Derek Tucker, Ph.D.
Statistician

Dept. 5573 |  Statistical Sciences



Office 505.284.8415

Mobile 505.379.0572

Web research.tetonedge.net



Sandia National Laboratories



sandia.gov<http://www.sandia.gov/>




________________________________
From: Kiran Vaddi ***@***.***>
Sent: Thursday, July 7, 2022 5:20:32 PM
To: jdtuck/fdasrsf_python ***@***.***>
Cc: Tucker, James Derek ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] [jdtuck/fdasrsf_python] Warping function between a pair of open curves (Discussion #29)


Hi @jdtuck<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjdtuck&data=05%7C01%7Cjdtuck%40sandia.gov%7Cdb698ae259744b5c03fb08da606f4b74%7C7ccb5a20a303498cb0c129007381b574%7C1%7C0%7C637928328383872043%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Y6E1CBWU9WZzxA%2BIebXsmma2gifLU%2BUI0RqA7gGSjYA%3D&reserved=0>,
I have a quick question regarding computing the warping function between a pair of open curves.
I am trying to align the solid black curve to the solid blue curve shown below but the result after the optimization (dotted black curve) doesn't move much away from the solid black curve.
[image]<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F48496966%2F177886354-7d992a08-4678-4a80-9cb4-7a7ebd8b9337.png&data=05%7C01%7Cjdtuck%40sandia.gov%7Cdb698ae259744b5c03fb08da606f4b74%7C7ccb5a20a303498cb0c129007381b574%7C1%7C0%7C637928328383872043%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=22Z1jHoQSgQqB%2FByn1w9rrkQNkaWo1ou649Bpt2jlDc%3D&reserved=0>

I am doing this using the following code:

beta2best, q2best, Rbest, gamIbest = fs.find_rotation_and_seed_coord(f1, f2, closed=0, rotation=False, method="DP")

where f1, f2 are of the shape (2, 100) and I am plotting beta2best as the warped dotted black curve above.

In reality, these can be aligned using the time_warping module but that seems to only align the major peak (see the plot below) of the two curves while I am particularly interested in the full diffeomorphism (i.e. the gamma function that optimally matches one curve to other by stretching and bending) that aligns the two curves.

[image]<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F48496966%2F177886998-0df8398f-e0ad-4f65-9aad-47b193e1ab9a.png&data=05%7C01%7Cjdtuck%40sandia.gov%7Cdb698ae259744b5c03fb08da606f4b74%7C7ccb5a20a303498cb0c129007381b574%7C1%7C0%7C637928328383872043%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=u2uPqFbWheql7LxMLod%2BP8Ap9m1%2Fx2ZE%2BL3zXuFcOao%3D&reserved=0>

Do you have any suggestions?

TIA
Kiran Vaddi

**Reply by kiranvad** [reply]
Correct, I thought so too. I think what I am looking for is some form of mapping that fully deforms one curve.
Essentially, adding all the peaks necessary to make the blue curve look like the black one. 
I thought of using the curves module after reading Ch 5 from the Functional and Shape Data Analysis book. 

I was hoping to match the two curves like the bottle and heart symbol example shown below. Maybe I am missing something here?

![image](https://user-images.githubusercontent.com/48496966/177891946-9863c709-8272-44c2-b846-57cab73eb71e.png)

**Reply by jdtuck** [reply]
It is not clear what you are wanting to do, are you working with curves in $\mathbb{R}^n$ or are you working with functions ($\mathbb{R}^1$)? The package can align each, but are different functions.

**Reply by kiranvad** [reply]
It's the latter and I misunderstood what the gamma function was doing in principle. Your previous answer clarified it.
